### PR TITLE
Paginate solutions - reduce memory consumption

### PR DESCRIPTION
**Pull request by Hettomei**
You may not love the 'style' since I do not use Rails magic to create/use object. 
But I cannot find ways to 'not load all entries' when using rails models

# What have been done : 

- Move leaderboard in it's own object
- Create ShowChallenge for various information
- create Solutions to query uniq solution
- paginate solutions (but always display ALL solutions)
- fix pagination : 
  - when not connected
  - when connected with 0 contribution
  - when connected with >= 1 contribution
- fix 'position' of a solution (need to query the 'offset')
- Add tests for queries

# Additionnal infos : 

About 'reduce memory consumption' I use : 

```
pp NewRelic::Agent::Samplers::MemorySampler.new.sampler.get_sample
```

At the beginning and at the end of 'ChallengeController#show'
It looks like it seems close to what heroku returns.

with 
- "Simple, Practical, and Common - 18089 entries" 
- as the best player (display all solution)
- after 10 reload page

Before this PR :
beginning of #show 136, end of #show 191.578125
-> I get lot of requestTimeout, and MemorySampler varie from 110 to 210 

With this PR : 
beginning of #show 136, end of #show 136
-> no requestTimeout, and MemorySampler is stable

### CODE REVIEW

File: app/services/solution.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    @users = users
+    @position = position
+  end
+  attr_reader :solution, :position
+
+  def id
+    solution[:entry_id]
+  end
+
+  def score
```
Comment (igrigorik) [inline app/services/solution.rb]: Minor, but you can refactor (score, script, comments, created_at) as dynamically defined methods. 

```[x,y,z].each { define_method(...)}```

File: app/views/challenges/show.erb [hunk 2]
```diff
@@ -1,90 +1,73 @@
-<% challenge_id(@challenge) %>
+<% challenge_id(@show_challenge) %>
```
Comment (igrigorik) [inline app/views/challenges/show.erb]: Any reason why we can't keep existing @challenge? It's not clear what `show_` adds in terms of readability.

File: app/services/show_challenge.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+require_relative '../repositories/repository_challenge'
+require 'forwardable'
+
+# retieve specific data for show page
+# avoid loading all 'entries' to limit memory
+class ShowChallenge
```
Comment (igrigorik) [inline app/services/show_challenge.rb]: Can't say I'm in love with the name, but struggling to come up with better alternatives.. Any others that you were considering?

File: app/services/solutions.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
+require_relative '../repositories/repository_challenge'
+require_relative './solution'
+
+class Solutions
```
Comment (igrigorik) [inline app/services/solutions.rb]: `Solution` vs `Solutions` is a subtle and easy to miss difference. Perhaps `Solution` and `Entries` or `Submissions` instead?

**Comment by igrigorik** [comment]
@Hettomei this is a big one, kudos for taking it on. Left a few nitpick suggestions above. 

What's left, modulo fixing the failing build?

**Comment by Hettomei** [comment]
> What's left

Implement the system that display solutions step by step (based on the score of players)
Cureently I display all

File: app/services/solution.rb (hunk 1 above)
Comment (Hettomei) [reply app/services/solution.rb]: For me it is not a refactor, it is a code golf . And as somebody told one day 'I do greppable code'. I think it s better to keep 'def something' .

File: app/services/solutions.rb (hunk 4 above)
Comment (Hettomei) [reply app/services/solutions.rb]: Nice.
I ll try Submissions && Solution

File: app/views/challenges/show.erb (hunk 2 above)
Comment (Hettomei) [reply app/views/challenges/show.erb]: It was to show to other devs that it is not a 'Challenge model' but 'service/show_challenge' to avoid confusion.

But I agree with you 'ShowChallenge' was a bad name, and so this @show_challenge

**Comment by Hettomei** [comment]
> Implement the system that display solutions step by step (based on the score of players)
--> done

> What's left
- 'count remaining solution'
- add offset to display real solution position
- fix tests
- add tests

**Comment by Hettomei** [comment]
Hi
thanks for the reminder.

I have a working solution on another branch (paginate-solution-2), but I don't like it: 
- too much mongodb query 
- hard to maintain for only one page
- far away from rails style/logic
- zero test (I found edge case that I didn't understood at first, so I should add lot of test to show what I found)

I have an idea to simplify the number/difference/maintenability of queries but I need to take time for that. And I will have time at the end of july

**Comment by igrigorik** [comment]
@Hettomei 👍

**Comment by Hettomei** [comment]
updated description : only test left

File: app/services/solutions.rb (hunk 4 above)
Comment (igrigorik) [reply app/services/solutions.rb]: @Hettomei bump on this one. Could we rename this to Submissions?

File: app/services/solutions.rb [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    @solutions ||= RepositoryChallenge.solutions(
+      challenge_id: challenge_id,
+      min_score: visible_score,
+      per_page: PER_PAGE,
+      page: page
+    )
+  end
+
+  def visible_score
+    @visible_score ||= if !player
```
Comment (igrigorik) [inline app/services/solutions.rb]: Could we unroll this one into multi-line? Easier to grok :)

File: app/views/shared/_entries.erb [hunk 6]
```diff
@@ -0,0 +1,42 @@
+<% @solutions.each do |solution, user, position| %>
```
Comment (igrigorik) [inline app/views/shared/_entries.erb]: if we're renaming to submissions, rename the template as well to _submissions.erb?

**Comment by Hettomei** [comment]
About perf :
In prod env, after 'warming up' the page
I will push what rails gives (ms to display a view)
I will push memory used.

I don't know how to do other good benchmark.

About the review : thanks, i'll update asap.

**Comment by Hettomei** [comment]
# same benchmark, but in a table

|                                  | ram                | page view rails | note                 |
|----------------------------------|--------------------|-----------------|----------------------|
| dev: master; worst player        | 250..256           | OK in 3078ms    |                      |
| dev: pull request; worst player  | 134 then grow 0.26 | OK in 1077ms    | :1st_place_medal:    |
|                                  |                    |                 |                      |
| dev: master; best player         | 272..286           | OK in 7287ms    | sometimes crash :-1: |
| dev: pull request; best player   | 134 then grow 0.26 | OK in 1192ms    | :1st_place_medal:    |
|                                  |                    |                 |                      |
| prod: master; worst player       | 220..238           | OK in 3139ms    |                      |
| prod: pull request; worst player | 116                | OK in 921ms     | :1st_place_medal:    |
|                                  |                    |                 |                      |
| prod: master; best player        | 220..238           | OK in 6682ms    |                      |
| prod: pull request; best player  | 120                | OK in 1147ms    | :1st_place_medal:    |

**Comment by Hettomei** [comment]
Interpretation of the benchmark : 
No need to go deeper, it seems that this PR improve vimgolf.

File: app/views/shared/_entries.erb (hunk 6 above)
Comment (Hettomei) [reply app/views/shared/_entries.erb]: done

File: app/services/solutions.rb (hunk 5 above)
Comment (Hettomei) [reply app/services/solutions.rb]: done

**Comment by Hettomei** [comment]
Again, ready to merge from my point of view ;)

**Comment by igrigorik** [comment]
Awesome work sir. Merging.
